File: continuity_receipt/anchor.py
"""OpenTimestamps proof verification — 0.3 anchor companion tool.

Scope (ANCHORING.md §5): replay a detached ``.ots`` proof from the file
digest to every attestation, and verify a Bitcoin attestation against an
80-byte block header by exact byte equality with the header's merkle root
(bytes 36..68). Header *chain* validation (proof-of-work, confirmation
depth, reorg handling) is out of scope by design: supply the header from a
source you trust (Bitcoin Core, a header database) and read
``status=verified`` as "this digest is the merkle root of that header",
not "that header is buried under N blocks".

Wire format: python-opentimestamps ``DetachedTimestampFile`` — magic,
uint8 major version, file-hash op tag, digest, then the timestamp tree.
Varints are LEB128 (unsigned little-endian base-128), *not* Bitcoin
CompactSize; the distinction is pinned by tests/test_anchor.py.

Statuses:
  verified   — a Bitcoin attestation's reached digest equals the supplied
               header's merkle root
  unverified — the proof parses and binds to the digest, but reaches only
               pending/unknown attestations, or a Bitcoin attestation
               without a header to check it against
  mismatch   — the proof does not bind to the expected digest, the supplied
               header's merkle root does not match, or the bundle anchor
               entry disagrees
  invalid    — malformed, truncated, or uses an unsupported operation
"""

from __future__ import annotations

import hashlib
import json
import sys
from dataclasses import dataclass, field
from typing import Any

MAGIC = b"\x00OpenTimestamps\x00\x00Proof\x00\xbf\x89\xe2\xe8\x84\xe8\x92\x94"
MAJOR_VERSION = 1

MAX_MSG = 4096
MAX_RESULT = 4096
MAX_ATT_PAYLOAD = 8192
MAX_URI = 1000
MAX_NODES = 10000
MAX_DEPTH = 256
MAX_VARUINT_BYTES = 10

HASH_OPS = {
    0x02: ("sha1", 20),
    0x03: ("ripemd160", 20),
    0x08: ("sha256", 32),
}
BINARY_OPS = {0xF0: "append", 0xF1: "prepend"}
UNARY_OPS = {0xF2: "reverse", 0xF3: "hexlify"}
KECCAK_TAG = 0x67

ATT_PENDING = bytes.fromhex("83dfe30d2ef90c8e")
ATT_BITCOIN = bytes.fromhex("0588960d73d71901")
ATT_LITECOIN = bytes.fromhex("06869a0d73d71b45")

HEADER_LEN = 80
MERKLE_OFFSET = 36
TIME_OFFSET = 68


class AnchorError(Exception):
    """Proof-format error; ``code`` is a machine-readable identifier."""

    def __init__(self, code: str, detail: str):
        super().__init__(detail)
        self.code = code
        self.detail = detail


class _Reader:
    def __init__(self, data: bytes):
        self.data = data
        self.pos = 0

    def read(self, n: int) -> bytes:
        if n < 0 or self.pos + n > len(self.data):
            raise AnchorError("truncated", "proof is truncated")
        out = self.data[self.pos : self.pos + n]
        self.pos += n
        return out

    def read_varuint(self) -> int:
        value = 0
        shift = 0
        for _ in range(MAX_VARUINT_BYTES):
            b = self.read(1)[0]
            value |= (b & 0x7F) << shift
            if not (b & 0x80):
                return value
            shift += 7
        raise AnchorError("invalid_varuint", "varuint exceeds 10 bytes")

    def read_varbytes(self, max_len: int, min_len: int = 0) -> bytes:
        length = self.read_varuint()
        if length > max_len:
            raise AnchorError("invalid_length", f"varbytes length {length} exceeds {max_len}")
        if length < min_len:
            raise AnchorError("invalid_length", f"varbytes length {length} below {min_len}")
        return self.read(length)

    def assert_eof(self) -> None:
        if self.pos != len(self.data):
            raise AnchorError("trailing_data", "unexpected bytes after proof node")


@dataclass(frozen=True)
class Attestation:
    kind: str  # "pending" | "bitcoin" | "litecoin" | "unknown"
    value: Any

    def as_dict(self) -> dict:
        if self.kind == "unknown":
            return {"kind": self.kind, **self.value}
        return {"kind": self.kind, "value": self.value}


@dataclass
class ParsedProof:
    version: int
    file_hash_op: str
    file_digest: bytes
    leaves: list  # list[(reached_digest: bytes, Attestation)]


@dataclass
class AnchorResult:
    status: str
    code: str
    detail: str = ""
    file_digest: str = ""
    attestations: list = field(default_factory=list)
    confirmed: dict | None = None

    @property
    def ok(self) -> bool:
        return self.status == "verified"

    def as_dict(self) -> dict:
        out = {
            "status": self.status,
            "code": self.code,
            "detail": self.detail,
            "file_digest": self.file_digest,
            "attestations": self.attestations,
        }
        if self.confirmed is not None:
            out["confirmed"] = self.confirmed
        return out


def _digest_bytes(value: Any) -> bytes | None:
    """Normalize bytes / hex / ``sha256:<hex>`` into a raw digest."""
    if isinstance(value, (bytes, bytearray)):
        return bytes(value)
    if not isinstance(value, str):
        return None
    text = value.strip()
    if ":" in text:
        prefix, _, rest = text.partition(":")
        if prefix != "sha256":
            return None
        text = rest
    if text.startswith("0x"):
        text = text[2:]
    try:
        return bytes.fromhex(text)
    except ValueError:
        return None


def _apply_op(name: str, arg: bytes | None, msg: bytes) -> bytes:
    if len(msg) > MAX_MSG:
        raise AnchorError("message_too_long", f"operation input exceeds {MAX_MSG} bytes")
    if name == "append":
        result = msg + arg
    elif name == "prepend":
        result = arg + msg
    elif name == "reverse":
        result = msg[::-1]
    elif name == "hexlify":
        result = msg.hex().encode()
    else:
        try:
            result = hashlib.new(name, msg).digest()
        except ValueError as exc:  # pragma: no cover - host-dependent
            raise AnchorError("unsupported_op", f"hash {name} unavailable: {exc}") from exc
    if not result or len(result) > MAX_RESULT:
        raise AnchorError("invalid_result", f"{name} produced {len(result)} bytes")
    return result


def _parse_attestation(payload_tag: bytes, payload: bytes) -> Attestation:
    reader = _Reader(payload)
    if payload_tag == ATT_PENDING:
        uri = reader.read_varbytes(MAX_URI)
        reader.assert_eof()
        return Attestation("pending", uri.decode("utf-8", errors="replace"))
    if payload_tag in (ATT_BITCOIN, ATT_LITECOIN):
        height = reader.read_varuint()
        reader.assert_eof()
        return Attestation("bitcoin" if payload_tag == ATT_BITCOIN else "litecoin", height)
    return Attestation("unknown", {"tag": payload_tag.hex(), "payload": payload.hex()})


def _parse_timestamp(
    reader: _Reader, msg: bytes, leaves: list, state: dict, depth: int
) -> None:
    if depth > MAX_DEPTH:
        raise AnchorError("recursion_limit", "timestamp tree exceeds depth limit")
    state["nodes"] += 1
    if state["nodes"] > MAX_NODES:
        raise AnchorError("too_many_nodes", "timestamp tree exceeds node limit")

    def handle(tag: bytes) -> None:
        if tag == b"\x00":
            tag8 = reader.read(8)
            payload = reader.read_varbytes(MAX_ATT_PAYLOAD)
            leaves.append((msg, _parse_attestation(tag8, payload)))
            return
        tag_byte = tag[0]
        if tag_byte == KECCAK_TAG:
            raise AnchorError("unsupported_op", "keccak256 proofs are not supported")
        if tag_byte in HASH_OPS:
            name, _ = HASH_OPS[tag_byte]
            _parse_timestamp(reader, _apply_op(name, None, msg), leaves, state, depth + 1)
            return
        if tag_byte in BINARY_OPS:
            arg = reader.read_varbytes(MAX_RESULT, min_len=1)
            _parse_timestamp(reader, _apply_op(BINARY_OPS[tag_byte], arg, msg), leaves, state, depth + 1)
            return
        if tag_byte in UNARY_OPS:
            _parse_timestamp(reader, _apply_op(UNARY_OPS[tag_byte], None, msg), leaves, state, depth + 1)
            return
        raise AnchorError("unknown_op", f"unknown operation tag 0x{tag_byte:02x}")

    tag = reader.read(1)
    while tag == b"\xff":
        handle(reader.read(1))
        tag = reader.read(1)
    handle(tag)


def parse_detached(data: bytes) -> ParsedProof:
    reader = _Reader(data)
    if reader.read(len(MAGIC)) != MAGIC:
        raise AnchorError("bad_magic", "not an OpenTimestamps detached proof")
    version = reader.read(1)[0]
    if version != MAJOR_VERSION:
        raise AnchorError("unsupported_version", f"major version {version} not supported")
    hash_tag = reader.read(1)[0]
    if hash_tag not in HASH_OPS:
        raise AnchorError("unknown_op", f"unsupported file hash op 0x{hash_tag:02x}")
    file_hash_op, digest_len = HASH_OPS[hash_tag]
    file_digest = reader.read(digest_len)
    leaves: list = []
    _parse_timestamp(reader, file_digest, leaves, {"nodes": 0}, 0)
    reader.assert_eof()
    return ParsedProof(version, file_hash_op, file_digest, leaves)


def verify_proof(
    data: bytes,
    expected_digest: Any,
    block_headers: dict | None = None,
) -> AnchorResult:
    """Verify a detached OTS proof against a digest and optional headers.

    ``block_headers`` maps Bitcoin block height -> 80-byte header. When
    provided, a matching-height Bitcoin attestation is checked by merkle
    root equality; other heights are reported but not verified.
    """
    expected = _digest_bytes(expected_digest)
    if expected is None:
        return AnchorResult("invalid", "bad_digest", "expected digest is not bytes/hex/sha256:<hex>")

    try:
        proof = parse_detached(data)
    except AnchorError as exc:
        return AnchorResult("invalid", exc.code, exc.detail)

    file_digest = f"{proof.file_hash_op}:{proof.file_digest.hex()}"
    if len(expected) != len(proof.file_digest) or expected != proof.file_digest:
        return AnchorResult(
            "mismatch",
            "digest_mismatch",
            "proof was not created for the expected digest",
            file_digest=file_digest,
        )

    headers: dict[int, bytes] = {}
    if block_headers:
        for height, header in block_headers.items():
            if isinstance(header, str):
                header = _digest_bytes(header) or b""
            if not isinstance(header, (bytes, bytearray)) or len(header) != HEADER_LEN:
                return AnchorResult("invalid", "bad_header", f"header for height {height} is not 80 bytes")
            headers[int(height)] = bytes(header)

    attestations = [att.as_dict() for _, att in proof.leaves]
    bitcoins = [(msg, att) for msg, att in proof.leaves if att.kind == "bitcoin"]

    for msg, att in bitcoins:
        header = headers.get(att.value)
        if header is None:
            continue
        if msg == header[MERKLE_OFFSET : MERKLE_OFFSET + 32]:
            confirmed = {
                "block_height": att.value,
                "merkle_root": msg.hex(),
                "header_time": int.from_bytes(header[TIME_OFFSET : TIME_OFFSET + 4], "little"),
            }
            return AnchorResult(
                "verified",
                "anchor_verified",
                f"digest is the merkle root of Bitcoin block {att.value}",
                file_digest=file_digest,
                attestations=attestations,
                confirmed=confirmed,
            )

    if bitcoins and headers:
        heights = ", ".join(str(att.value) for _, att in bitcoins)
        return AnchorResult(
            "mismatch",
            "header_mismatch",
            f"supplied header(s) do not match the proof (attested height(s): {heights})",
            file_digest=file_digest,
            attestations=attestations,
        )

    if bitcoins:
        return AnchorResult(
            "unverified",
            "anchor_unverified",
            "proof reaches a Bitcoin attestation; supply the 80-byte block header to confirm",
            file_digest=file_digest,
            attestations=attestations,
        )

    kinds = {att.kind for _, att in proof.leaves}
    if "pending" in kinds:
        return AnchorResult(
            "unverified",
            "anchor_pending",
            "proof reaches only pending calendar attestations; upgrade it after confirmation",
            file_digest=file_digest,
            attestations=attestations,
        )
    if "litecoin" in kinds:
        return AnchorResult(
            "unverified",
            "anchor_unverified",
            "proof reaches a Litecoin attestation (not checked by this tool)",
            file_digest=file_digest,
            attestations=attestations,
        )
    if proof.leaves:
        return AnchorResult(
            "unverified",
            "anchor_unknown_type",
            "proof reaches only unknown attestation types",
            file_digest=file_digest,
            attestations=attestations,
        )
    return AnchorResult("invalid", "anchor_empty", "proof contains no attestations", file_digest=file_digest)


def _bundle_digest(bundle_path: str, target: str) -> tuple[str, dict | None]:
    with open(bundle_path, "r", encoding="utf-8") as handle:
        bundle = json.load(handle)
    from .bundle import receipt_digest

    for receipt in bundle.get("receipts", []):
        if isinstance(receipt, dict) and receipt.get("receipt_id") == target:
            binding = None
            for anchor in bundle.get("anchors", []) or []:
                if isinstance(anchor, dict) and anchor.get("target") == target:
                    meta = anchor.get("anchor")
                    if isinstance(meta, dict) and meta.get("type") == "opentimestamps":
                        binding = anchor
                        break
            return receipt_digest(receipt), binding
    raise AnchorError("unknown_target", f"no receipt {target!r} in bundle")


def main(argv=None) -> int:
    import argparse

    parser = argparse.ArgumentParser(
        prog="continuity-receipt-anchor",
        description="Verify an OpenTimestamps (.ots) proof against a receipt digest.",
    )
    sub = parser.add_subparsers(dest="command", required=True)
    verify = sub.add_parser("verify", help="verify a detached .ots proof")
    verify.add_argument("proof", help="path to the detached .ots proof")
    source = verify.add_mutually_exclusive_group(required=True)
    source.add_argument("--digest", help="expected digest: hex or sha256:<hex>")
    source.add_argument("--bundle", help="bundle JSON; use with --target")
    verify.add_argument("--target", help="receipt_id inside the bundle (with --bundle)")
    verify.add_argument("--header", help="80-byte Bitcoin block header, hex")
    verify.add_argument("--height", type=int, help="block height the header belongs to")
    verify.add_argument("--json", action="store_true", help="print the full result as JSON")

    args = parser.parse_args(argv)

    try:
        expected = args.digest
        binding = None
        if args.bundle:
            if not args.target:
                parser.error("--bundle requires --target")
            expected, binding = _bundle_digest(args.bundle, args.target)
    except AnchorError as exc:
        print(json.dumps({"status": "invalid", "code": exc.code, "detail": exc.detail}))
        return 1
    except OSError as exc:
        print(json.dumps({"status": "invalid", "code": "io_error", "detail": str(exc)}))
        return 1

    headers = None
    if args.header:
        headers = {args.height if args.height is not None else 0: args.header}

    try:
        with open(args.proof, "rb") as handle:
            data = handle.read()
    except OSError as exc:
        print(json.dumps({"status": "invalid", "code": "io_error", "detail": str(exc)}))
        return 1

    result = verify_proof(data, expected, block_headers=headers)

    if binding is not None:
        from .bundle import receipt_digest

        if binding.get("hash") != expected:
            result = AnchorResult(
                "mismatch",
                "anchor_binding_mismatch",
                "bundle anchor hash disagrees with the receipt digest",
                file_digest=result.file_digest,
                attestations=result.attestations,
            )
        else:
            result.detail += "; bundle binding ok"

    if args.json:
        print(json.dumps(result.as_dict(), indent=2))
    else:
        print(f"{result.status}: {result.code} — {result.detail}")
        if result.file_digest:
            print(f"  file digest: {result.file_digest}")
        for att in result.attestations:
            print(f"  attestation: {att}")
        if result.confirmed:
            print(f"  confirmed: {result.confirmed}")
    return 0 if result.ok else 1


if __name__ == "__main__":
    sys.exit(main())

File: rust/tests/anchor.rs
//! Anchor companion tests: synthetic proofs plus the real `.ots` fixtures
//! under `../vectors/anchor/` (mirrors `tests/test_anchor.py`).

use std::fs;
use std::path::PathBuf;

use continuity_receipt::anchor::{parse_detached, verify_proof, verify_proof_raw, Attestation};
use serde_json::{json, Map, Value};
use sha2::{Digest, Sha256};

const ATT_PENDING: [u8; 8] = [0x83, 0xdf, 0xe3, 0x0d, 0x2e, 0xf9, 0x0c, 0x8e];
const ATT_BITCOIN: [u8; 8] = [0x05, 0x88, 0x96, 0x0d, 0x73, 0xd7, 0x19, 0x01];
const MAGIC: &[u8] = b"\x00OpenTimestamps\x00\x00Proof\x00\xbf\x89\xe2\xe8\x84\xe8\x92\x94";

fn varuint(mut value: u64) -> Vec<u8> {
    let mut out = Vec::new();
    loop {
        let mut byte = (value & 0x7F) as u8;
        value >>= 7;
        if value != 0 {
            byte |= 0x80;
            out.push(byte);
        } else {
            out.push(byte);
            return out;
        }
    }
}

fn varbytes(bytes: &[u8]) -> Vec<u8> {
    let mut out = varuint(bytes.len() as u64);
    out.extend_from_slice(bytes);
    out
}

fn pending_att(uri: &str) -> Vec<u8> {
    let mut out = ATT_PENDING.to_vec();
    out.extend_from_slice(&varbytes(&varbytes(uri.as_bytes())));
    out
}

fn bitcoin_att(height: u64) -> Vec<u8> {
    let mut out = ATT_BITCOIN.to_vec();
    out.extend_from_slice(&varbytes(&varuint(height)));
    out
}

fn detached(digest: &[u8], tree: &[u8]) -> Vec<u8> {
    let mut out = MAGIC.to_vec();
    out.push(1);
    out.push(0x08);
    out.extend_from_slice(digest);
    out.extend_from_slice(tree);
    out
}

fn header_for(merkle: &[u8], time: u32) -> Vec<u8> {
    assert_eq!(merkle.len(), 32);
    let mut out = vec![0x02, 0x00, 0x00, 0x00];
    out.extend_from_slice(&[0x11; 32]);
    out.extend_from_slice(merkle);
    out.extend_from_slice(&time.to_le_bytes());
    out.extend_from_slice(&[0x00; 8]);
    out
}

fn repo_root() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .parent()
        .expect("rust crate sits inside the repository")
        .to_path_buf()
}

fn fixture(name: &str) -> Vec<u8> {
    fs::read(repo_root().join("vectors").join("anchor").join(name))
        .unwrap_or_else(|error| panic!("cannot read fixture {name}: {error}"))
}

fn fixture_headers() -> Map<String, Value> {
    let text = fs::read_to_string(repo_root().join("vectors").join("anchor").join("headers.json"))
        .expect("headers.json");
    let data: Value = serde_json::from_str(&text).expect("headers JSON");
    data["headers"].as_object().expect("headers object").clone()
}

fn hex(bytes: &[u8]) -> String {
    bytes.iter().map(|byte| format!("{byte:02x}")).collect()
}

struct Sample {
    digest: Vec<u8>,
    message: Vec<u8>,
    proof: Vec<u8>,
}

fn sample() -> Sample {
    let digest = Sha256::digest(b"continuity-receipt test vector").to_vec();
    let message = Sha256::digest(&digest).to_vec();
    let mut tree = vec![0x08, 0x00];
    tree.extend_from_slice(&bitcoin_att(800_000));
    let proof = detached(&digest, &tree);
    Sample {
        digest,
        message,
        proof,
    }
}

#[test]
fn bitcoin_attestation_verified_against_header() {
    let sample = sample();
    let mut headers = std::collections::BTreeMap::new();
    headers.insert(800_000, header_for(&sample.message, 1_700_000_000));
    let result = verify_proof_raw(&sample.proof, &sample.digest, &headers);
    assert_eq!(result.status, "verified");
    assert_eq!(result.code, "anchor_verified");
    assert!(result.ok());
    let confirmed = result.confirmed.expect("confirmed");
    assert_eq!(confirmed["block_height"], json!(800_000));
    assert_eq!(confirmed["header_time"], json!(1_700_000_000u32));
    assert_eq!(confirmed["merkle_root"], json!(hex(&sample.message)));
}

#[test]
fn bitcoin_attestation_without_header_is_unverified() {
    let sample = sample();
    let result = verify_proof_raw(&sample.proof, &sample.digest, &Default::default());
    assert_eq!(result.status, "unverified");
    assert_eq!(result.code, "anchor_unverified");
    assert_eq!(result.attestations[0]["kind"], json!("bitcoin"));
}

#[test]
fn header_mismatch() {
    let sample = sample();
    let mut headers = std::collections::BTreeMap::new();
    headers.insert(800_000, header_for(&[0u8; 32], 1_700_000_000));
    let result = verify_proof_raw(&sample.proof, &sample.digest, &headers);
    assert_eq!(result.status, "mismatch");
    assert_eq!(result.code, "header_mismatch");
}

#[test]
fn digest_mismatch() {
    let sample = sample();
    let other = Sha256::digest(b"other").to_vec();
    let result = verify_proof_raw(&sample.proof, &other, &Default::default());
    assert_eq!(result.status, "mismatch");
    assert_eq!(result.code, "digest_mismatch");
}

#[test]
fn pending_and_unknown_attestations() {
    let sample = sample();
    let unknown_tag: [u8; 8] = [0x00, 0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77];
    let mut tree = vec![0xFF, 0xF0];
    tree.extend_from_slice(&varbytes(b"AAAA"));
    tree.push(0x00);
    tree.extend_from_slice(&pending_att(
        "https://alice.btc.calendar.opentimestamps.org",
    ));
    tree.push(0xF0);
    tree.extend_from_slice(&varbytes(b"BBBB"));
    tree.push(0x00);
    tree.extend_from_slice(&unknown_tag);
    tree.extend_from_slice(&varbytes(&[0x01, 0x02, 0x03]));

    let proof = detached(&sample.digest, &tree);
    let result = verify_proof_raw(&proof, &sample.digest, &Default::default());
    assert_eq!(result.status, "unverified");
    assert_eq!(result.code, "anchor_pending");
    let kinds: std::collections::BTreeSet<&str> = result
        .attestations
        .iter()
        .filter_map(|attestation| attestation["kind"].as_str())
        .collect();
    assert_eq!(kinds, ["pending", "unknown"].into_iter().collect());
}

#[test]
fn leb128_varuints_not_compact_size() {
    let sample = sample();
    let uri = "a".repeat(128);
    let mut tree = vec![0x00];
    tree.extend_from_slice(&ATT_PENDING);
    tree.extend_from_slice(&[0x82, 0x01]);
    tree.extend_from_slice(&[0x80, 0x01]);
    tree.extend_from_slice(uri.as_bytes());
    let parsed = parse_detached(&detached(&sample.digest, &tree)).expect("parses");
    match &parsed.leaves[0].1 {
        Attestation::Pending(value) => assert_eq!(value, &uri),
        other => panic!("expected pending, got {other:?}"),
    }

    let mut tree300 = vec![0x00];
    tree300.extend_from_slice(&ATT_BITCOIN);
    tree300.extend_from_slice(&varbytes(&[0xAC, 0x02]));
    let parsed = parse_detached(&detached(&sample.digest, &tree300)).expect("parses");
    assert_eq!(parsed.leaves[0].1, Attestation::Bitcoin(300));
}

#[test]
fn reverse_and_prepend_ops() {
    let sample = sample();
    let prefix = vec![b'X'; 8];
    let mut tree = vec![0xF1];
    tree.extend_from_slice(&varbytes(&prefix));
    tree.push(0xF2);
    tree.push(0x00);
    tree.extend_from_slice(&pending_att("https://ots.example"));
    let parsed = parse_detached(&detached(&sample.digest, &tree)).expect("parses");
    let (reached, attestation) = &parsed.leaves[0];
    let mut expected = prefix.clone();
    expected.extend_from_slice(&sample.digest);
    expected.reverse();
    assert_eq!(reached, &expected);
    assert!(matches!(attestation, Attestation::Pending(_)));
}

#[test]
fn truncated_and_bad_magic() {
    let sample = sample();
    let truncated = &sample.proof[..sample.proof.len() - 3];
    let result = verify_proof_raw(truncated, &sample.digest, &Default::default());
    assert_eq!(result.code, "truncated");

    let mut bad = vec![0u8; 16];
    bad.extend_from_slice(&sample.proof);
    let result = verify_proof_raw(&bad, &sample.digest, &Default::default());
    assert_eq!(result.code, "bad_magic");
}

#[test]
fn keccak_is_unsupported_not_silent() {
    let sample = sample();
    let mut tree = vec![0x67, 0x00];
    tree.extend_from_slice(&pending_att("https://ots.example"));
    let proof = detached(&sample.digest, &tree);
    let error = parse_detached(&proof).expect_err("keccak is unsupported");
    assert_eq!(error.code, "unsupported_op");
    let result = verify_proof_raw(&proof, &sample.digest, &Default::default());
    assert_eq!(result.status, "invalid");
    assert_eq!(result.code, "unsupported_op");
}

#[test]
fn overlong_varuint_is_rejected() {
    let sample = sample();
    let mut tree = vec![0x00];
    tree.extend_from_slice(&ATT_PENDING);
    tree.extend_from_slice(&[0x80; 10]);
    tree.push(0x00);
    let error = parse_detached(&detached(&sample.digest, &tree)).expect_err("varuint too long");
    assert_eq!(error.code, "invalid_varuint");
}

#[test]
fn hello_world_binds_to_the_file_digest() {
    let parsed = parse_detached(&fixture("hello-world.txt.ots")).expect("parses");
    let expected = Sha256::digest(fixture("hello-world.txt")).to_vec();
    assert_eq!(parsed.file_digest, expected);
    assert_eq!(parsed.file_hash_op, "sha256");
}

#[test]
fn hello_world_verified_against_block_358391() {
    let proof = fixture("hello-world.txt.ots");
    let expected = Sha256::digest(fixture("hello-world.txt")).to_vec();
    let result = verify_proof(&proof, &json!(hex(&expected)), Some(&fixture_headers()));
    assert_eq!(result.status, "verified");
    assert_eq!(result.confirmed.expect("confirmed")["block_height"], json!(358391));
}

#[test]
fn sha1_bitcoin_pdf_verified_against_block_465751() {
    let proof = fixture("bitcoin.pdf.ots");
    let parsed = parse_detached(&proof).expect("parses");
    assert_eq!(parsed.file_hash_op, "sha1");
    let result = verify_proof(
        &proof,
        &json!(hex(&parsed.file_digest)),
        Some(&fixture_headers()),
    );
    assert_eq!(result.status, "verified");
    assert_eq!(result.confirmed.expect("confirmed")["block_height"], json!(465751));
}

#[test]
fn gdp_2025_proof_verified_against_block_912095() {
    let proof = fixture("gdp2q25-2nd.pdf.ots");
    let parsed = parse_detached(&proof).expect("parses");
    let result = verify_proof(
        &proof,
        &json!(hex(&parsed.file_digest)),
        Some(&fixture_headers()),
    );
    assert_eq!(result.status, "verified");
    assert_eq!(result.confirmed.expect("confirmed")["block_height"], json!(912095));
}

#[test]
fn known_and_unknown_notary_reports_both() {
    let proof = fixture("known-and-unknown-notary.txt.ots");
    let parsed = parse_detached(&proof).expect("parses");
    let result = verify_proof(&proof, &json!(hex(&parsed.file_digest)), None);
    assert_eq!(result.status, "unverified");
    assert_eq!(result.code, "anchor_pending");
    let kinds: std::collections::BTreeSet<&str> = result
        .attestations
        .iter()
        .filter_map(|attestation| attestation["kind"].as_str())
        .collect();
    assert_eq!(kinds, ["pending", "unknown"].into_iter().collect());
}

#[test]
fn different_blockchains_fails_loudly_on_keccak() {
    let proof = fixture("different-blockchains.txt.ots");
    let error = parse_detached(&proof).expect_err("keccak path");
    assert_eq!(error.code, "unsupported_op");
    let result = verify_proof(&proof, &json!(hex(&[0u8; 32])), None);
    assert_eq!(result.status, "invalid");
    assert_eq!(result.code, "unsupported_op");
}
